topology/probes/blockdev/blockdev.go

```diff
@@ -0,0 +1,462 @@
+		return
+	}
+
+	if p.findBlockDev(path, blockID, newDevInfo) == false {
+		p.unregisterBlockdev(path)
+	}
+
+}
+
+func (p *ProbeHandler) registerBlockdev(blockdev BlockDevice, parentWWN string) *graph.Node {
+
+	var groupNode *graph.Node
+	childCount := len(blockdev.Children)
+	if childCount > 0 {
+		for i := range blockdev.Children {
+			groupNode = p.registerBlockdev(blockdev.Children[i], blockdev.WWN)
+		}
+		if blockdev.Type == multipathType {
+			groupNode = p.addGroup(blockdev, parentWWN)
+		}
+	} else {
+		groupNode = p.addGroup(blockdev, parentWWN)
+	}
+
+	p.Lock()
+	defer p.Unlock()
+	var graphMetaData graph.Metadata
+
+	if _, ok := p.blockdevMap[blockdev.Name]; ok {
+		return groupNode
+	}
+
+	p.Ctx.Graph.Lock()
+	defer p.Ctx.Graph.Unlock()
+
+	node := p.Ctx.Graph.LookupFirstNode(graph.Metadata{"Path": p.getPath(blockdev)})
+
+	if node == nil {
+		graphMetaData = p.getMetaData(blockdev, childCount, parentWWN)
+		var err error
+		if node, err = p.Ctx.Graph.NewNode(graph.GenID(), graphMetaData); err != nil {
+			p.Ctx.Logger.Error(err)
+			return nil
+		}
+
+		// If there is a WWN - check to see if there are other paths to the same block device.
+		if blockdev.WWN != "" {
+			multiPathNodes := p.Ctx.Graph.GetNodes(graph.Metadata{"BlockID": p.getID(blockdev)})
+
+			for _, multiPathNode := range multiPathNodes {
+				topology.AddLink(p.Ctx.Graph, multiPathNode, node, "multipath", nil)
+			}
+		}
+
+		topology.AddOwnershipLink(p.Ctx.Graph, groupNode, node, nil)
+
+		// Link physical disks and DVD/rom devices to the blockdevGroupName
+		if blockdev.Type == "disk" || blockdev.Type == "rom" {
+			topology.AddLink(p.Ctx.Graph, p.Groups[blockdevGroupName], node, "connected", nil)
+		}
+
+	}
+
+	for i := range blockdev.Children {
+		childNode := p.Ctx.Graph.LookupFirstNode(graph.Metadata{"Name": p.getName(blockdev.Children[i])})
+		if childNode != nil {
+			topology.AddLink(p.Ctx.Graph, childNode, node, "connected", nil)
+		}
+	}
+
+	p.blockdevMap[blockdev.Name] = blockdevInfo{
+		ID:   blockdev.Name,
+		Node: node,
+	}
+	return groupNode
+}
+
+func (p *ProbeHandler) unregisterBlockdev(id string) {
+	p.Lock()
+	defer p.Unlock()
+
+	info, ok := p.blockdevMap[id]
+	if !ok {
+		return
+	}
+
+	p.Ctx.Graph.Lock()
+	if err := p.Ctx.Graph.DelNode(info.Node); err != nil {
+		p.Ctx.Graph.Unlock()
+		p.Ctx.Logger.Error(err)
+		return
+	}
+	p.Ctx.Graph.Unlock()
+
+	delete(p.blockdevMap, id)
+}
+
+func (p *ProbeHandler) connect() error {
+	var (
+		cmdOut []byte
+		err    error
+		result Devices
+	)
+
+	testFile := os.Getenv("SKYDIVE_BLOCKDEV_TEST_FILE")
+
+	if testFile != "" {
+		if cmdOut, err = exec.Command("cat", testFile).Output(); err != nil {
+			p.Ctx.Logger.Error(err)
+			os.Exit(1)
+		}
+	} else {
+		if cmdOut, err = exec.Command("lsblk", "-pO", "--json").Output(); err != nil {
+			p.Ctx.Logger.Error(err)
+			os.Exit(1)
+		}
+	}
+
+	jsonBlockdevs := string(cmdOut[:])
+	if err = json.Unmarshal([]byte(jsonBlockdevs), &result); err != nil {
```

Comment:
json.Unmarshal(cmdOut, ....
is not working ? instead of passing via a string

Reply:
if err = json.Unmarshal([]byte(cmdOut[:]), &result); err != nil {

Reply:
Ok.